assembl/models/discussion.py

```diff
@@ -659,33 +673,45 @@ def decorate_instance(
             & (LocalUserRole.requested == False)),
         backref="participant_in_discussion")
 
-    def get_participants_query(self, ids_only=False, include_readers=False):
+    def get_participants_query(self, ids_only=False, include_readers=False, current_user=None):
         from .auth import AgentProfile, LocalUserRole
         from .generic import Content
         from .post import Post
         from .action import ViewPost
         from .idea_content_link import Extract
-
-        query = self.db.query(AgentProfile.id).join(LocalUserRole,
-                LocalUserRole.user_id==AgentProfile.id).filter(
-            LocalUserRole.discussion_id == self.id).union(
-            self.db.query(AgentProfile.id).join(Post,
-                Post.creator_id==AgentProfile.id).filter(
-            Post.discussion_id == self.id)).union(
-            self.db.query(AgentProfile.id).join(
-                Extract, Extract.creator_id==AgentProfile.id).filter(
-            Extract.discussion_id == self.id))
-        query = query.union(self.db.query(AgentProfile.id).join(UserRole,
-                UserRole.user_id==AgentProfile.id))
+        from .announcement import Announcement
+        from .attachment import Attachment
+        post = with_polymorphic(Post, [Post])
+        attachment = with_polymorphic(Attachment, [Attachment])
+        extract = with_polymorphic(Extract, [Extract])
+        db = self.db
+        queries = [
+            db.query(LocalUserRole.user_id.label('user_id')).filter(
+                LocalUserRole.discussion_id == self.id),
+            db.query(post.creator_id.label('user_id')).filter(
+                post.discussion_id == self.id),
+            db.query(extract.creator_id.label('user_id')).filter(
+                extract.discussion_id == self.id),
+            db.query(extract.owner_id.label('user_id')).filter(
+                extract.discussion_id == self.id),
+            db.query(Announcement.creator_id.label('user_id')).filter(
+                Announcement.discussion_id == self.id),
+            db.query(attachment.creator_id.label('user_id')).filter(
+                attachment.discussion_id == self.id),
+            db.query(UserRole.user_id.label('user_id')),
+        ]
+        if self.creator_id is not None:
+            queries.append(db.query(literal(self.creator_id).label('user_id')))
+        if current_user is not None:
+            queries.append(db.query(literal(current_user).label('user_id')))
         if include_readers:
-            query = query.union(
-                self.db.query(ViewPost.actor_id).join(
+            queries.append(db.query(ViewPost.actor_id.label('user_id')).join(
                 Content, Content.id==ViewPost.post_id).filter(
                 Content.discussion_id==self.id))
-        query = query.distinct()
+        query = queries[0].union(*queries[1:]).distinct()
```

Comment:
Interesting. Wonder what the performance of this is? (Does it even matter, it's just for one API set?)

Reply:
It's actually not that bad, and much better than getting all users in the systems when we actually fetch the data for those users (remember that api2 does not optimize joinloads!)